packages/ember/tests/routing/query_params_test/shared_state_test.js

```diff
@@ -0,0 +1,88 @@
+import {
+  Controller,
+  Service
+} from 'ember-runtime';
+import Ember from 'ember';
+import { run } from 'ember-metal';
+import { jQuery } from 'ember-views';
+import { QueryParamTestCase, moduleFor } from 'internal-test-helpers';
+
+class SharedStateQPTestCase extends QueryParamTestCase {
+  boot() {
+    this.setupApplication();
+    return this.visitApplication();
+  }
+}
+
+moduleFor('Query Params - shared service state', class extends SharedStateQPTestCase {
+  setupApplication() {
+    this.router.map(function() {
+      this.route('home', { path: '/' });
+      this.route('dashboard');
+    });
+
+    this.application.register('service:filters', Service.extend({
+      shared: true
+    }));
+
+    this.registerController('home', Controller.extend({
+      filters: Ember.inject.service(),
+      queryParams: [
+        { 'filters.shared': 'shared' }
+      ]
+    }));
+
+    this.registerController('dashboard', Controller.extend({
+      filters: Ember.inject.service(),
+      queryParams: [
+        { 'filters.shared': 'shared' }
+      ]
+    }));
+
+    this.registerTemplate('application', `{{link-to 'Home' 'home' }} <div> {{outlet}} </div>`);
+    this.registerTemplate('home', `{{link-to 'Dashboard' 'dashboard' }}{{input type="checkbox" id='filters-checkbox' checked=(mut filters.shared) }}`);
+    this.registerTemplate('dashboard', `{{link-to 'Home' 'home' }}`);
+  }
+  visitApplication() {
+    return this.visit('/');
+  }
+
+  ['@test should not fail if shared state is modified from default before transition']() {
+    let assert = this.assert;
+    let done = assert.async();
```

Comment:
No need to use `done` when returning a Promise. QUnit will wait from the Promise to finish automatically.